File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys
input = sys.stdin.readline

def preorder(root):
    if root != '.':
        print(root, end='')
        preorder(graph[root][0])
        preorder(graph[root][1])

def inorder(root):
    if root != '.':
        inorder(graph[root][0])
        print(root, end='')
        inorder(graph[root][1])

def postorder(root):
    if root != '.':
        postorder(graph[root][0])
        postorder(graph[root][1])
        print(root, end='')

graph = dict()
n = int(input())
for _ in range(n):
    root, left, right = input().strip().split()
    graph[root] = (left, right)

preorder('A')
print()
inorder('A')
print()
postorder('A')

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        char value;
        Node left;
        Node right;

        Node(char value) {
            this.value = value;
        }
    }

    static void preorder(Node node) {
        if (node != null) {
            System.out.print(node.value);
            preorder(node.left);
            preorder(node.right);
        }
    }

    static void inorder(Node node) {
        if (node != null) {
            inorder(node.left);
            System.out.print(node.value);
            inorder(node.right);
        }
    }

    static void postorder(Node node) {
        if (node != null) {
            postorder(node.left);
            postorder(node.right);
            System.out.print(node.value);
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        Map<Character, Node> tree = new HashMap<>();
        StringTokenizer st;

        for (int i = 0; i < n; i++) {
            char node = (char) ('A' + i);
            tree.put(node, new Node(node));
        }

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            char parent = st.nextToken().charAt(0);
            char left = st.nextToken().charAt(0);
            char right = st.nextToken().charAt(0);

            Node parentNode = tree.get(parent);
            if (left != '.') {
                parentNode.left = tree.get(left);
            }
            if (right != '.') {
                parentNode.right = tree.get(right);
            }
        }
        Node root = tree.get('A');

        preorder(root);
        System.out.println();
        inorder(root);
        System.out.println();
        postorder(root);
    }
}
